File: app/src/main/java/cl/figonzal/lastquakechile/services/NightModeService.java
package cl.figonzal.lastquakechile.services;

import android.app.Activity;
import android.content.SharedPreferences;
import android.view.Window;

import androidx.annotation.NonNull;
import androidx.appcompat.app.AppCompatDelegate;
import androidx.lifecycle.Lifecycle;
import androidx.lifecycle.LifecycleObserver;
import androidx.lifecycle.OnLifecycleEvent;
import androidx.preference.PreferenceManager;

import cl.figonzal.lastquakechile.R;
import timber.log.Timber;

public class NightModeService implements LifecycleObserver {

    private final Activity activity;
    private final Window window;

    public NightModeService(Activity activity, @NonNull Lifecycle lifecycle, Window window) {
        this.activity = activity;
        this.window = window;

        lifecycle.addObserver(this);
    }

    /**
     * Funcion que permite revisar y establecer el modo noche desde Shared Preference Settings
     */
    @SuppressWarnings("unused")
    @OnLifecycleEvent(Lifecycle.Event.ON_START)
    public void checkNightMode() {

        //Leer preference settings
        SharedPreferences sharedPreferences = PreferenceManager.getDefaultSharedPreferences(activity);

        boolean manual_night_mode = sharedPreferences.getBoolean(activity.getString(R.string.NIGHT_MODE_MANUAL_KEY), false);

        boolean auto_night_mode = sharedPreferences.getBoolean(activity.getString(R.string.NIGHT_MODE_AUTO_KEY), false);

        //MODO MANUAL
        //Si el modo manual esta activado
        if (manual_night_mode) {

            AppCompatDelegate.setDefaultNightMode(AppCompatDelegate.MODE_NIGHT_YES);
            window.setStatusBarColor(activity.getColor(R.color.colorPrimaryVariantNightMode));

            //fixAdViewNightMode(activity);

            Timber.i(activity.getString(R.string.TAG_NIGHT_MODE_MANUAL) + ": " + activity.getString(R.string.TAG_NIGHT_MODE_STATUS_ON));
        }

        //MODO AUTOMATICO
        //Si el modo automatico esta activado
        else if (auto_night_mode) {

            AppCompatDelegate.setDefaultNightMode(AppCompatDelegate.MODE_NIGHT_AUTO_BATTERY);

            Timber.i(activity.getString(R.string.TAG_NIGHT_MODE_AUTO) + ": " + activity.getString(R.string.TAG_NIGHT_MODE_STATUS_ON));
        }

        //DESACTIVADO
        //Si el modo nocturno esta desactivado
        else {

            AppCompatDelegate.setDefaultNightMode(AppCompatDelegate.MODE_NIGHT_NO);
            window.setStatusBarColor(activity.getColor(R.color.colorPrimaryVariant));

            Timber.i(activity.getString(R.string.TAG_NIGHT_MODE) + ": " + activity.getString(R.string.TAG_NIGHT_MODE_STATUS_OFF));
        }
    }
}

File: app/src/main/java/cl/figonzal/lastquakechile/services/NightModeService.kt
package cl.figonzal.lastquakechile.services

import android.app.Activity
import androidx.appcompat.app.AppCompatDelegate
import androidx.lifecycle.DefaultLifecycleObserver
import androidx.lifecycle.LifecycleOwner
import androidx.preference.PreferenceManager
import cl.figonzal.lastquakechile.R
import timber.log.Timber

class NightModeService(
    private val activity: Activity
) : DefaultLifecycleObserver {


    override fun onStart(owner: LifecycleOwner) {
        super.onStart(owner)

        checkNightMode()
    }

    private fun checkNightMode() {

        //Leer preference settings
        val sharedPreferences = PreferenceManager.getDefaultSharedPreferences(activity)

        val manualNightMode =
            sharedPreferences.getBoolean(activity.getString(R.string.NIGHT_MODE_KEY), false)

        //MANUAL MODE
        //manual mode activated
        when {
            manualNightMode -> {

                AppCompatDelegate.setDefaultNightMode(AppCompatDelegate.MODE_NIGHT_YES)

                Timber.i(
                    activity.getString(R.string.TAG_NIGHT_MODE) + ": ON"
                )
            }

            else -> {
                AppCompatDelegate.setDefaultNightMode(AppCompatDelegate.MODE_NIGHT_NO)

                Timber.i(
                    activity.getString(R.string.TAG_NIGHT_MODE) + ": OFF"
                )
            }
        }
    }
}